### ISSUE TITLE
Automated style checking/linting

### ISSUE THREAD

**Issue by janden** [kymatio/kymatio#626]
To speed up PRs, it would be good to have a check in the CI that verifies the code style. From what I understand, there are two basic tools that can help here, `pycodestyle` and `autopep8`. The former checks whether the code follows the PEP 8 standard, while the latter can help reformat the code to be compliant. Ideally, we'd have `pycodestyle` run in one of the CIs (which?) and if it fails, instruct the author to run `autopep8`.

The crux here is that these tools don't quite work out of the box. For one, we may want to follow certain parts of PEP 8 while not others (line break before or after binary operator? maximum line length?). We can start with something loose, then make it stricter later on if needed. Another issue is that `autopep8` won't fix all the errors. You can make it fix more by specifying `--aggressive`, but that sounds a little risky. Part of the issue, I think, is that it's not clear that all errors can be fixed automatically and may require human judgment.

If I run `autopep8` on the current dev (942d8b0159bf538372f5a48aa0d8d8176af6416b) and then `pycodestyle kymatio --ignore=W504,E501` (ignore line break after binary operator and long lines), I get the output
```
kymatio/scattering2d/filter_bank.py:42:42: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:42:44: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:42:46: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:42:49: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:43:65: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:54:44: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/filter_bank.py:96:13: E741 ambiguous variable name 'l'
kymatio/scattering2d/frontend/base_frontend.py:27:92: E226 missing whitespace around arithmetic operator
kymatio/scattering2d/frontend/numpy_frontend.py:34:12: E713 test for membership should be 'not in'
kymatio/scattering2d/frontend/tensorflow_frontend.py:37:16: E713 test for membership should be 'not in'
kymatio/scattering2d/frontend/torch_frontend.py:97:12: E713 test for membership should be 'not in'
kymatio/scattering3d/filter_bank.py:13:38: E741 ambiguous variable name 'O'
kymatio/scattering3d/filter_bank.py:39:9: E741 ambiguous variable name 'l'
kymatio/scattering3d/filter_bank.py:41:25: E226 missing whitespace around arithmetic operator
kymatio/scattering3d/filter_bank.py:49:32: E741 ambiguous variable name 'O'
kymatio/scattering3d/filter_bank.py:78:23: E741 ambiguous variable name 'O'
kymatio/scattering3d/filter_bank.py:117:29: E741 ambiguous variable name 'O'
kymatio/scattering3d/filter_bank.py:117:39: E741 ambiguous variable name 'l'
kymatio/scattering3d/filter_bank.py:144:29: E226 missing whitespace around arithmetic operator
kymatio/scattering3d/filter_bank.py:144:31: E226 missing whitespace around arithmetic operator
kymatio/scattering3d/filter_bank.py:162:8: E741 ambiguous variable name 'l'
kymatio/scattering3d/utils.py:29:14: E741 ambiguous variable name 'O'
kymatio/scattering3d/core/scattering3d.py:30:9: E741 ambiguous variable name 'l'
kymatio/scattering3d/frontend/base_frontend.py:23:30: E741 ambiguous variable name 'O'
kymatio/scattering3d/frontend/numpy_frontend.py:32:17: W503 line break before binary operator
kymatio/scattering3d/frontend/tensorflow_frontend.py:48:16: E713 test for membership should be 'not in'
kymatio/scattering3d/frontend/torch_frontend.py:55:17: W503 line break before binary operator
kymatio/scattering3d/frontend/torch_frontend.py:72:12: E713 test for membership should be 'not in'
kymatio/scattering1d/frontend/numpy_frontend.py:30:12: E713 test for membership should be 'not in'
kymatio/scattering1d/frontend/tensorflow_frontend.py:31:12: E713 test for membership should be 'not in'
kymatio/scattering1d/frontend/torch_frontend.py:84:12: E713 test for membership should be 'not in'
kymatio/scattering1d/backend/torch_skcuda_backend.py:197:35: E226 missing whitespace around arithmetic operator
```
These are definitely fixable by hand. Alternatively, we can ignore them for now (like E741, although it's probably a good idea to fix) and add them in later.

This doesn't look at other potential problems, like import ordering, unused variables, and so on. For that, we could run `pylint` and `isort`, but I haven't looked at those yet.

**Comment by edouardoyallon** [comment]
this sounds like very helpful

**Comment by lostanlen** [comment]
I agree. Another worth checking out it black

https://www.youtube.com/watch?v=ia19n_yK4Qs

**Comment by janden** [comment]
Yeah I've had good experience with black. Will look into this.

**Comment by eickenberg** [comment]
Yes, it would be great to agree on a format moving forward. Once we agree on that format, we need to retro-correct the rest of the codebase swiftly.

I looked at black at some point and it was the most reasonable and far-sighted approach I had seen so far. There might be some compatibility issues with pep8 and such here and there, but as long as we are using the same thing and require style to be correct before PR merges, this doesn't matter.

**Comment by janden** [comment]
Black is largely compatible with pep8, right? The most important thing here is that the changes be done while the number of PRs is small. Otherwise, we'll have a lot of conflicts that have to be rebased. In other words, this is something that we need to set up before or after the sprint, not during. Before might be easier, since we'll have a bunch of trailing PRs. That also means clearing up the PRs we have right now, which is mostly the work from this summer. Some we can merge but some we may want to close for now.

The other thing to consider is how to make sure this is checked automatically before merging PRs and (very important) how devs can check this locally before submitting a PR.

**Comment by lostanlen** [comment]
> The other thing to consider is how to make sure this is checked automatically before merging PRs and (very important) how devs can check this locally before submitting a PR.

Is there a way to config the check so that it only looks at the modified/added lines of the PR and not the entirety of the file? If so, we could do only the latter part of the plan (check that all PRs pass the black test) and this way, little by little, our code will become entirely black-compliant.
